### ISSUE TITLE
Custom djangorestframework-api-key header not passed in schema requests

### ISSUE THREAD

**Issue by stefanofusai** [tfranzel/drf-spectacular#1102]
The issue I'm experiencing is related to authorizing an API key to view the schema.
I have different types of API keys subclassing [djangorestframework-api-key](https://florimondmanca.github.io/djangorestframework-api-key/), each with access to different (or all) endpoints.

What I would like to do is to show different paths on my schema, based on what API key the user is using.
I solved this by setting `SERVE_PUBLIC: False` - everything works fine except for the **Authorize** button.

As an unauthorized user visits the Swagger UI page, no paths are shown (which is the expected behavior)
However, once I click **Authorize** and input my API key, the schema doesn't refresh.
Looking into the Chrome console, I can see that a request is in fact made to fetch the new schema, but it doesn't include the required header that would allow to fetch the available endpoints.

If I manually set the header in my headers, everything works fine - meaning that the issue relies indeed in the fact that the refresh request sent after clicking authorize doesn't include the header.
I would also like to point out that "Try it out" requests made after authorizing the API key do include the header.

So, to recap: everything works perfectly except for the schema request made after clicking authorizing, which doesn't include the required header.

Here are my settings for reference:

```py
SPECTACULAR_SETTINGS = {
    "SCHEMA_PATH_PREFIX": ...,
    "SERVE_PUBLIC": False,
    "SERVE_INCLUDE_SCHEMA": False,
    "APPEND_COMPONENTS": {
        "securitySchemes": {
            "ApiKeyAuth": {"type": "apiKey", "in": "header", "name": "X-API-Key"}
        }
    },
    "SECURITY": [{"ApiKeyAuth": []}],
    "AUTHENTICATION_WHITELIST": ["rest_framework_api_key.permissions.HasAPIKey"],
    "TITLE": ...,
    "DESCRIPTION": ...,
    "CONTACT": ...,
    "VERSION": ...,
    "SERVERS": ...,
    "SWAGGER_UI_SETTINGS": {"persistAuthorization": True},
}
```

Thanks for the help!

**Comment by tfranzel** [comment]
Hi! interesting.... so `djangorestframework-api-key` is a hacky library which bypasses the regular auth mechanisms of DRF. That is why it is in the blueprint section, as there is no possible clean integration of it. 

That said, there is some magic going on in the js templates and spectacular only considers proper auth methods of the schema view here. Any other auth method would have worked out of the box.

You need to slightly patch the schema view to make this work. Your custom auth name simply has to be handed through, so the frontend knows when to inject the header:

```python
from drf_spectacular.views import SpectacularSwaggerView
from drf_spectacular.settings import spectacular_settings

class CustomSpectacularSwaggerView(SpectacularSwaggerView):
    def _get_schema_auth_names(self):
        return super()._get_schema_auth_names() + [
            list(auth.keys())[0] for auth in spectacular_settings.SECURITY
        ]
```

use this view instead of the regular one and it should work.

**Comment by stefanofusai** [comment]
Works perfectly! Thank you so much for the quick help

**Comment by stefanofusai** [comment]
Hey, reopening this as I've noticed that the `/schema/` request is made without the input API Key - the paths dict in the output yml file is empty since no authorization is provided.
Not sure if this can be fixed or not - thank you regardless for the help

**Comment by tfranzel** [comment]
> Hey, reopening this as I've noticed that the /schema/ request is made without the input API Key - the paths dict in the output yml file is empty since no authorization is provided

so it wasn't working before? once you click authorize, the schema should be re-fetched with the added header.

here is where the token gets injected (given the `ApiKeyAuth` name is returned in the list from the patched `CustomSpectacularSwaggerView`):

https://github.com/tfranzel/drf-spectacular/blob/fbbcab362d40cdd6df2f34f66c04d81dcf821fe8/drf_spectacular/templates/drf_spectacular/swagger_ui.js#L86

**Comment by stefanofusai** [comment]
Sorry for my bad explanation. The "main" schema request works perfectly and it solved my original issue.
The schema request I'm talking about is the one made when clicking the schema URL at the top of the page, right under the title.

**Comment by tfranzel** [comment]
well that is likely impossible to fix since it merely a `href` the browser follows. Adding the header is an manual thing. This would work in case of cookie auth, since they are automatically attached by the browser.

I could come up with a piggy-backed query param to the schema endpoint that auths the request in parallel to the existing method but it is hacky and non-trivial to get right. I would prefer not to support risky stuff and so I would ask you to find out yourself how to do it exactly.

**Comment by stefanofusai** [comment]
That's the same answer I came up with in my mind right after submitting my reply.
Thanks for the help once again!
